File: src/backtracking/LeetCode131.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function (s) {
  let result = [],
    path = [],
    subString = ''

  const backTracking = function (_s, startIndex) {
    // 终止条件
    if (startIndex >= _s.length) {
      result.push(path.slice())
      return
    }
    // 单层搜索逻辑
    for (let i = startIndex; i < _s.length; i++) {
      // 是否回文子串
      subString = _s.slice(startIndex, i + 1)
      if (!isPalindrome(subString)) {
        continue
      }
      path.push(subString)
      // 递归
      backTracking(_s, i + 1)
      // 回溯
      path.pop()
    }
  }

  // 判断是否是回文子串
  const isPalindrome = function (s) {
    let left = 0,
      right = s.length - 1
    while (left < right) {
      if (s[left] != s[right]) {
        return false
      }
      left++
      right--
    }
    return true
  }

  backTracking(s, 0)
  return result
}

File: src/backtracking/LeetCode131.java
package backtracking;

import java.util.ArrayList;
import java.util.List;

/**
 * @author: Mr.Ye
 * @description: LeetCode131.分割回文串(https://leetcode.cn/problems/palindrome-partitioning/description/)——中等
 * @createDate: 2024/12/01
 */
public class LeetCode131 {
  private List<List<String>> result = new ArrayList<>();
  private List<String> path = new ArrayList<>();

  public List<List<String>> partition(String s) {
    backTracking(s, 0);
    return result;
  }

  private void backTracking(String s, int startIndex) {
    String substring = "";
    // 终止条件
    if (startIndex >= s.length()) {
      result.add(new ArrayList<>(path));
      return;
    }
    // 单层搜索逻辑
    for (int i = startIndex; i < s.length(); i++) {
      // 切割线：startIndex，子串范围：[startIndex, i]
      // 获取子串
      substring = s.substring(startIndex, i + 1);
      // 非回文子串，则继续搜索
      if (!isPalindrome(substring)) {
        continue;
      }
      path.add(substring);
      // 递归
      backTracking(s, i + 1);
      // 回溯
      path.remove(path.size() - 1);
    }
  }

  private boolean isPalindrome(String s) {
    int left = 0, right = s.length() - 1;
    while (left < right) {
      if (s.charAt(left) != s.charAt(right)) {
        return false;
      }
      left++;
      right--;
    }
    return true;
  }
}
